### PR TITLE
inventory aws_ec2 - assume role using iam_role_arn parameter to describe regions

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

This is a proposal to solve #566 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

aws_ec2

### CODE REVIEW

**Comment by abikouo** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Review by alinabuzachis — COMMENTED** [review]
@abikouo Thank you for taking time to work on this. LGTM. I was just wondering if you could add a unit test.

**Comment by abikouo** [comment]
> @abikouo Thank you for taking time to work on this. LGTM. I was just wondering if you could add a unit test.

Hi @alinabuzachis this is not a good candidate for unit testing as this will need to create a lot of mock objects to simply capture an exception, I would rather go for integration testing but I am missing authorization to create an IAM role, for the time being, this can be merged like this.

**Comment by abikouo** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Comment by alinabuzachis** [comment]
@abikouo I guess there is an issue with the requirements for units in general.

 ```
2022-02-23 09:24:51.345399 \| controller \| ==================================== ERRORS ====================================
--
1965 | 2022-02-23 09:24:51.345403 \| controller \| ________________________ ERROR collecting test session _________________________
1966 | 2022-02-23 09:24:51.345407 \| controller \| ImportError while importing test module '/usr/share/ansible/collections/ansible_collections/amazon/aws/tests/unit/plugins/modules/test_ec2_vpc_dhcp_option.py'.
1967 | 2022-02-23 09:24:51.345411 \| controller \| Hint: make sure your test modules/packages have valid Python names.
1968 | 2022-02-23 09:24:51.345414 \| controller \| Traceback:
1969 | 2022-02-23 09:24:51.345417 \| controller \| /usr/lib64/python3.8/importlib/__init__.py:127: in import_module
1970 | 2022-02-23 09:24:51.345420 \| controller \|     return _bootstrap._gcd_import(name[level:], package, level)
1971 | 2022-02-23 09:24:51.345424 \| controller \| /tmp/ansible-test-fppp5phd/ansible/utils/collection_loader/_collection_finder.py:402: in load_module
1972 | 2022-02-23 09:24:51.345427 \| controller \|     exec(code_obj, module.__dict__)
1973 | 2022-02-23 09:24:51.345430 \| controller \| tests/unit/plugins/modules/test_ec2_vpc_dhcp_option.py:10: in <module>
1974 | 2022-02-23 09:24:51.345433 \| controller \|     from ansible_collections.amazon.aws.tests.unit.utils.amazon_placebo_fixtures import placeboify  # pylint: disable=unused-import
1975 | 2022-02-23 09:24:51.345437 \| controller \| E   ImportError: cannot import name 'placeboify' from 'ansible_collections.amazon.aws.tests.unit.utils.amazon_placebo_fixtures' (/usr/share/ansible/collections/ansible_collections/amazon/aws/tests/unit/utils/amazon_placebo_fixtures.py)
1976 | 2022-02-23 09:24:51.345440 \| controller \| - generated xml file: /usr/share/ansible/collections/ansible_collections/amazon/aws/tests/output/junit/python3.8-modules-units.xml -
1977 | 2022-02-23 09:24:51.345443 \| controller \| =========================== short test summary info ============================
1978 | 2022-02-23 09:24:51.345447 \| controller \| SKIPPED [1] tests/unit/plugins/modules/test_aws_s3.py:16: could not import 'boto3': No module named 'urllib3.packages.six'
1979 | 2022-02-23 09:24:51.345450 \| controller \| SKIPPED [1] tests/unit/utils/amazon_placebo_fixtures.py:10: could not import 'boto3': No module named 'urllib3.packages.six'
1980 | 2022-02-23 09:24:51.345454 \| controller \| ERROR ../../../../../../local/lib/python3.8/site-packages/ansible_test/_data
1981 | 2022-02-23 09:24:51.345457 \| controller \| ==================== 3 passed, 2 skipped, 1 error in 0.60s =====================
1982 | 2022-02-23 09:24:51.378895 \| controller \| ERROR: Command "pytest --forked -r a -n auto --color no -p no:cacheprovider -c /usr/local/lib/python3.8/site-packages/ansible_test/_data/pytest.ini --junit-xml /usr/share/ansible/collections/ansible_collections/amazon/aws/tests/output/junit/python3.8-modules-units.xml --strict-markers tests/unit/plugins/modules/test_aws_s3.py tests/unit/plugins/modules/test_cloudformation.py tests/unit/plugins/modules/test_ec2_group.py tests/unit/plugins/modules/test_ec2_vpc_dhcp_option.py" returned exit status 1.
1983 | 2022-02-23 09:24:52.009133 \| controller \| ERROR
1984 | 2022-02-23 09:24:52.009385 \| controller \| {
1985 | 2022-02-23 09:24:52.009424 \| controller \|   "delta": "0:00:01.556021",
1986 | 2022-02-23 09:24:52.009453 \| controller \|   "end": "2022-02-23 09:24:51.480333",
1987 | 2022-02-23 09:24:52.009479 \| controller \|   "msg": "non-zero return code",
1988 | 2022-02-23 09:24:52.009505 \| controller \|   "rc": 1,
1989 | 2022-02-23 09:24:52.009531 \| controller \|   "start": "2022-02-23 09:24:49.924312"
1990 | 2022-02-23 09:24:52.009555 \| controller \| }

```

**Review by alinabuzachis — APPROVED** [review]

File: changelogs/fragments/619-inventory-aws_ec2-assume-arn-role-when-listing-regions.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+- >-
+  aws_ec2 inventory - use the iam_role_arn configuration parameter to assume the role before trying to call DescribeRegions
+  if the regions configuration is not set and AWS credentials provided without enough privilege to perform the DescribeRegions action.
+  (https://github.com/ansible-collections/amazon.aws/issues/566)
```
Comment (markuman) [inline changelogs/fragments/619-inventory-aws_ec2-assume-arn-role-when-listing-regions.yml:5]:
```suggestion
bugfixes:
- >-
  aws_ec2 inventory - use the iam_role_arn configuration parameter to assume the role before trying to call DescribeRegions
  if the regions configuration is not set and AWS credentials provided without enough privilege to perform the DescribeRegions action.
  (https://github.com/ansible-collections/amazon.aws/issues/566).
```

in my understanding, this is a bugfix.

**Review by markuman — APPROVED** [review]

**Pull request merged**